src/csharp/Grpc.Core/Internal/GrpcThreadPool.cs

```diff
@@ -165,11 +170,19 @@ private void RunHandlerLoop(CompletionQueueSafeHandle cq, IProfiler optionalProf
                     try
                     {
                         var callback = cq.CompletionRegistry.Extract(tag);
-                        callback(success);
+                        // Use cached delegates to avoid unnecessary allocations
+                        if (!inlineHandlers)
+                        {
+                            ThreadPool.QueueUserWorkItem(success ? RunCompletionQueueEventCallbackSuccess : RunCompletionQueueEventCallbackFailure, callback);
```

Comment:
Have you measured the performance impact of this vs `Task.Run`? I'm not sure whether `ThreadPool.QueueUserWorkItem` uses the same task-stealing queues. But we could change it later in a non-breaking way, of course.